File: CSES/SearchAndSort/ConcertTicket.java
import java.util.*;
import java.io.*;

public class ConcertTicket {
    public static void main(String[] args) {
        FastReader sc = new FastReader();
        PrintWriter out = new PrintWriter(System.out);

        int n = sc.nextInt(); // Number of tickets
        int m = sc.nextInt(); // Number of customers

        // TreeMap to store Price -> Count (acts as a Multiset)
        TreeMap<Integer, Integer> tickets = new TreeMap<>();

        for (int i = 0; i < n; i++) {
            int price = sc.nextInt();
            tickets.put(price, tickets.getOrDefault(price, 0) + 1);
        }

        for (int i = 0; i < m; i++) {
            int maxPrice = sc.nextInt();
            
            // Find the largest price <= maxPrice
            Integer bestFit = tickets.floorKey(maxPrice);

            if (bestFit == null) {
                out.println("-1");
            } else {
                out.println(bestFit);
                
                // Remove one instance of this ticket
                int count = tickets.get(bestFit);
                if (count == 1) {
                    tickets.remove(bestFit);
                } else {
                    tickets.put(bestFit, count - 1);
                }
            }
        }
        out.close();
    }

    // Fast I/O Utility
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
    }
}

File: CSES/SearchAndSort/ConcertTicket.cpp
#include <iostream>
#include <set>
#include <vector>

using namespace std;

/**
 * Problem: Concert Tickets
 * Strategy: Use a multiset to store ticket prices. Use upper_bound 
 * to find the first price > maxPrice, then decrement the iterator 
 * to find the largest price <= maxPrice.
 * 
 * // using this as the java solution is TLE
 */

int main() {
    // Fast I/O: Equivalent to the FastReader in Java
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;

    multiset<int> tickets;
    for (int i = 0; i < n; i++) {
        int price;
        cin >> price;
        tickets.insert(price);
    }

    for (int i = 0; i < m; i++) {
        int maxPrice;
        cin >> maxPrice;

        // upper_bound returns an iterator to the first element STRICTLY GREATER than maxPrice
        auto it = tickets.upper_bound(maxPrice);

        // If the iterator is at the beginning, it means all tickets are > maxPrice
        if (it == tickets.begin()) {
            cout << -1 << "\n";
        } else {
            // Step back one position to get the largest element <= maxPrice
            --it;
            cout << *it << "\n";
            
            // Remove ONLY the specific instance pointed to by the iterator
            tickets.erase(it);
        }
    }

    return 0;
}